posts/2020/crud-pattern-in-react-implementation.md

```diff
@@ -0,0 +1,130 @@
+
+# BookForm Component - Implementation
+
+A form component creates or updates an entity on the server by performing API actions. In order to do this, it needs to fulfill following requirements:
+
+- Validate form and display validation errors (if any) of specific field
+- Display API response errors (if any) of specific field on the UI
+
+We can fulfill these requirements by using a `formFieldAttributes` object that defines validation rules for each field. We will check its details in the next section.
+
+## `formFieldAttributes` object
+
+It defines custom validation rules for every field in `{fieldName: { attr1: value1, ..., attrN: valueN }}` format. The rules can be specified as attributes as following :
+
+1. `key`: Used to generate and map error to the respective field
+2. `required`: Defines whether an empty field is allowed or not
+3. `customValidator()`: Provides custom validation logic for each field. Returns an error object in `{key: { error: true, helperText:"text" }}` format.
+
+```javascript
+const formFieldAttributes = {
+  title: {          // value of key ("title") is same as this (convention)
+    key: "title",   // Should be unique as it is used as field 'id'
+    required: true, // Whether an empty field is allowed or not
+    customValidator: value =>
+      value.length > 100 // Custom validation logic
+        ? { helperText: "enter less than 100 characters", error: true } // Object with error
+        : { helperText: "", error: false } // Object without error
+  },
+  ...
+}
+
+// Using 'key' as argument to fieldError() and fieldHelperText()
+// to map error and helperText
+<form id="bookForm">
+    <TextField
+        id={formFieldAttributes.title.key}
+        error={this.fieldError(formFieldAttributes.title.key)}
+        helperText={this.fieldHelperText(formFieldAttributes.title.key)}
+        ...
+    />
+...
+</form>
+
+```
+
+`formValidator()` function uses this object to validate form and generate errors (if any). We will check its functioning in the next section.
+
+## Form Validation (`formValidator()` function)
+
+```javascript
+const formValidator = (formFieldAttributes, userInput, isEditForm) => {
+  let formErrors = {};
+  for (let field in formFieldAttributes) {
+    let fieldObj = formFieldAttributes[field];
+    if (isEditForm && !fieldObj.editable) {
+      continue;
+    }
+    let fieldKey = fieldObj.key;
+
+    let value = userInput[fieldKey];
+
+    let fieldEmpty = !value;
+
+    if (fieldEmpty && fieldObj.required) {
+      formErrors[fieldKey] = {
+        error: true,
+        helperText: "this field should be non-empty"
+      };
+    } else if (fieldObj.customValidator && !fieldEmpty) {
+      let customValidator = fieldObj.customValidator;
+      let formError = customValidator(value);
+      if (formError.error) formErrors[fieldKey] = formError;
+    }
+  }
+  return { formErrors, errorCount: Object.keys(formErrors).length };
+};
+```
+
```

Comment:
We perhaps don't need the next para. So let's see - we can potentially delete this later. Because whatever next para describes is described above already.